source/extensions/filters/network/local_ratelimit/local_ratelimit.h

```diff
@@ -27,28 +28,66 @@ struct LocalRateLimitStats {
   ALL_LOCAL_RATE_LIMIT_STATS(GENERATE_COUNTER_STRUCT)
 };
 
+using LocalRateLimiterImplSharedPtr =
+    std::shared_ptr<Filters::Common::LocalRateLimit::LocalRateLimiterImpl>;
+
+class SharedRateLimitSingleton : public Singleton::Instance, Logger::Loggable<Logger::Id::filter> {
+public:
+  ~SharedRateLimitSingleton() override;
+
+  class Key : public std::pair<std::string, envoy::type::v3::TokenBucket> {
+  public:
+    using std::pair<std::string, envoy::type::v3::TokenBucket>::pair;
+
+    // Add absl::Hash support.
+    template <typename H>
+    friend H AbslHashValue(H h, const Key& key) { // NOLINT(readability-identifier-naming)
+      return H::combine(std::move(h), key.first, MessageUtil::hash(key.second));
+    }
+
+    bool operator==(const Key& that) const {
+      return (this->first == that.first) &&
+             Protobuf::util::MessageDifferencer::Equivalent(this->second, that.second);
+    }
+  };
+
+  std::pair<LocalRateLimiterImplSharedPtr, const Key*>
+  get(const envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit& proto_config,
+      std::function<LocalRateLimiterImplSharedPtr()> create_fn);
+
+  void removeIfUnused(const Key* key);
+
+private:
+  // Use node_hash_map so that a stable pointer to Key can be returned.
+  absl::node_hash_map<Key, std::weak_ptr<Filters::Common::LocalRateLimit::LocalRateLimiterImpl>>
+      limiters_;
+};
+
 /**
  * Configuration shared across all connections. Must be thread safe.
  */
 class Config : Logger::Loggable<Logger::Id::filter> {
 public:
   Config(
       const envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit& proto_config,
-      Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime);
+      Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime,
+      Singleton::Manager& singleton_manager);
+
+  ~Config();
 
   bool canCreateConnection();
   bool enabled() { return enabled_.enabled(); }
   LocalRateLimitStats& stats() { return stats_; }
 
 private:
   static LocalRateLimitStats generateStats(const std::string& prefix, Stats::Scope& scope);
-  void onFillTimer();
 
-  Filters::Common::LocalRateLimit::LocalRateLimiterImpl rate_limiter_;
+  LocalRateLimiterImplSharedPtr rate_limiter_;
   Runtime::FeatureFlag enabled_;
   LocalRateLimitStats stats_;
+  const SharedRateLimitSingleton::Key* shared_bucket_key_{};
+  std::shared_ptr<SharedRateLimitSingleton> shared_bucket_registry_;
 
-  std::vector<RateLimit::LocalDescriptor> descriptors_;
```

Comment:
This was previously unused, right? Was it there just to avoid a temporary allocation? If so you might want to leave it and make it const and add a small comment? Maybe I'm reading the diff wrong.

Reply:
It was always empty. But the function it is passed to takes an `absl::Span` not a `std::vector&`, so we don't need to construct a vector at all anymore; just pass an empty Span.